src/daft-recordbatch/src/ops/window.rs

```diff
@@ -69,10 +68,205 @@ impl RecordBatch {
+            // Always increment next_rank for regular rank()
+            if !dense {
+                next_rank += 1;
+            }
+
+            // Check if the current row has the same values as the previous row
+            let is_equal = comparator(i - 1, i);
+
+            if !is_equal {
+                // Different value, update rank
+                if dense {
+                    // For dense_rank, just increment by 1
+                    cur_rank += 1;
+                } else {
+                    // For rank(), use the next_rank which accounts for ties
+                    cur_rank = next_rank;
+                }
+            }
+
+            *rank = cur_rank;
+        }
+
+        let rank_series = UInt64Array::from((name.as_str(), rank_numbers)).into_series();
+        let rank_batch = Self::from_nonempty_columns(vec![rank_series])?;
+
+        self.union(&rank_batch)
+    }
+
+    // Helper method to process default values for window_offset
+    fn process_default_for_window(
+        &self,
+        default_expr: &ExprRef,
+        target_type: &DataType,
+        slice_start: usize,
+        slice_end: usize,
+        target_length: usize,
+    ) -> DaftResult<Series> {
+        // Only evaluate on the slice we need for efficiency
+        let default_slice = self.slice(slice_start, slice_end)?;
+        let def_col = default_slice.eval_expression(default_expr)?;
+
+        let def_col = if def_col.data_type() != target_type {
+            def_col.cast(target_type)?
+        } else {
+            def_col
+        };
+
+        if def_col.len() != target_length {
+            def_col.broadcast(target_length)
+        } else {
+            Ok(def_col)
+        }
+    }
+
+    pub fn window_offset(
+        &self,
+        name: String,
+        expr: ExprRef,
+        offset: i64,
+        default: Option<ExprRef>,
+    ) -> DaftResult<Self> {
+        // Short-circuit if offset is 0 - just return the value itself
+        if offset == 0 {
+            let expr_col = self.eval_expression(&expr)?;
+            let renamed_col = expr_col.rename(&name);
+            let result_batch = Self::from_nonempty_columns(vec![renamed_col])?;
+            return self.union(&result_batch);
+        }
+
+        let expr_col = self.eval_expression(&expr)?;
+        let abs_offset = offset.unsigned_abs() as usize;
+
+        let mut result_col = if self.is_empty() || abs_offset >= self.len() {
+            // Special case: empty array or offset exceeds array length
+            if let Some(default_expr) = default {
+                self.process_default_for_window(
+                    &default_expr,
+                    expr_col.data_type(),
+                    0,
+                    self.len(),
+                    self.len(),
+                )?
+            } else {
+                // Otherwise, create a column of nulls
+                Series::full_null(expr_col.name(), expr_col.data_type(), self.len())
+            }
```

Comment:
It looks like we're doing a lot of this `if let Some(default_expr) .. else` thing here. Maybe offload it into the helper function?